lesson21/src/js/main.js

```diff
@@ -0,0 +1,212 @@
+            class:"error-message"
+        },
+        str:error
+    });
+    element.appendChild(errorMessage);
+}
+
+const fetchData = async (endpointURL,element) => {
+    const response = await fetch(endpointURL);
+    if(!response.ok){
+        console.error(`${response.status}:${response.statusText}`);
+        createErrorMessage("Communication with the server is broken.",element);
+        return;
+    }
+    const json = await response.json();
+    return json.data;
+}
+
+const fetchContentsData = (endPointURL,element,ms) => new Promise(resolve => setTimeout(() => resolve(fetchData(endPointURL,element)),ms));
+
+const table = document.getElementById("js-table");
+
+const initUsersData = async () => {
+    loading.showLoading(table);
+    try{
+        const usersData = await fetchContentsData(endPointURL.usersData,table,500);
+        if (!usersData.length) {
+            createErrorMessage("No user.",table);
+            return;
+        }
+        renderTableElements(usersData);
+        sortTableBody(usersData);
+    }catch(error){
+        console.error(error);
+    }finally{
+       loading.removeLoading(table);
+    }
+}
+
+const renderTableElements = usersData => {
+    renderTableHeader();
+    renderTableBody(usersData);
+}
+
+const usersTableColumn = {
+    "id"     : "ID",
+    "name"   : "名前",
+    "gender" : "性別",
+    "age"    : "年齢"
+}
+
+const renderTableHeader = () => {
+    const thead = document.createElement("thead");
+    const tr = document.createElement("tr");
+    thead.className = "bg-slate-500";
+
+    for (const [userKey, userValue] of Object.entries(usersTableColumn)) {
+        const th = createAttributedElements({
+            tag:"th",
+            valuesByAttributes:{
+                class:"text-sm text-white px-6 py-4",
+            },
+            str:userValue
+        });
+
+        tr.appendChild(th);
+
+        addSortButtons(userKey,userValue,th);
+    };
+
+    table.appendChild(thead).appendChild(tr);
+}
+
+const renderTableBody = usersData => {
+    const tbody = document.createElement("tbody");
+    tbody.id = "js-usersTableBody";
+
+    for(const user of usersData){
+        const fragment = document.createDocumentFragment();
+        const tr = document.createElement("tr");
+
+        Object.keys(usersTableColumn).forEach((column) => {
+            const td = createAttributedElements({
+                tag:"td",
+                valuesByAttributes:{
+                    class:"border border-gray-300 px-4 py-2"
+                },
+                str:user[column]
+            })
+            fragment.appendChild(tr).appendChild(td);
+        });
+
+        table.appendChild(tbody).appendChild(fragment);
+    }
+}
+
+const sortButtons = [
+    {dataOrder:"asc",src:"../img/asc.svg", alt:"asc-image"},
+    {dataOrder:"desc",src:"../img/desc.svg",alt:"desc-image"},
+    {dataOrder:"both",src:"../img/both.svg",alt:"both-image"}
+]
+
+const sortByUsersTableColumn = {
+    "id" : "ID"
+}
```

Comment:
Q.
Does it have to be an Object here?
If we want to store the target category, we can use the
I thought an array would be fine.

Reply:
I used  the Object here,
because we thought they would be easier to manage when there are more elements to sort.

Initially, I had the following array in mind, 
but I thought it would become cumbersome to manage in the future when there are more elements to sort.

```JavaScript
const sortKeys = ["id"];
const sortValues = ["id"];
```

Reply:
keyの部分だけで十分な気がしてます。
valueも必要ですか？

例えば以下の様な配列すると、
わざわざkeyやvalueをObjectから取り出す必要もないです。
``` js

const sortCategories = ["id"];

```
はるさん(@haru-programming )の書いたcodesandboxを参考にしました:)

``` js
const renderTableHeader = () => {
 // emit
    if (sortCategories.includes(userKey)) {
      th.appendChild(createSortButtons(userKey));
    }
``` 

``` js

const sortTableBody = (usersData) => {
  // for (const key of Object.keys(sortByUsersTableColumn)) {
  sortCategories.forEach((category) => {
    const sortButtons = [
      ...document.getElementsByClassName(`js-${category}SortButton`),
    ];
    const sortButtonsBox = document.getElementById(`js-${category}Buttons-Box`);

    let dataOrder;
    sortButtonsBox.addEventListener("click", (e) => {
      const index = sortButtons.indexOf(e.target);
      dataOrder = e.target.getAttribute("data-order");

      if (e.currentTarget === e.target) {
        return;
      }

      toggleHiddenClassSortButton(sortButtons, index);
      sortUsersData(category, usersData, dataOrder);
    });
  });

  // }
};

```

Reply:
言われてみると、確かにkeyだけで十分なので、配列すれば良いと思いました。
コードまで書いていただき、ありがとうございます！
配列にする方法に、修正します。

Reply:
I changed the object to an array.
Thanks for all the advice.I learned a lot.
https://github.com/nakagawamai/js-lesson/pull/23/commits/9d278ab27770b13493036ee07e1b3ba1359c8132